jobs/jobs.go

```diff
@@ -1,35 +1,78 @@
 package jobs
 
-import "sync"
+import (
+	"context"
+	"sync"
+
+	"golang.org/x/sync/semaphore"
+)
 
 // Job represents a (long-running) process that can be spawned on a separate go-routine.
 // When encountering an error, a Job should send that error to the given channel and continue to run.
-type Job func(chan<- error)
+type Job func(ctx context.Context) error
 
 // JobManager manages multiple concurrent jobs and handles their errors. Can wait for all jobs and error handling to finish.
 type JobManager struct {
-	wgJobs  *sync.WaitGroup
-	wgErr   *sync.WaitGroup
-	errChan chan error
+	wgJobs      *sync.WaitGroup
+	done        chan struct{}
+	errChan     chan error
+	ctx         context.Context
+	once        *sync.Once
+	jobCapacity *capacityMgr
+}
+
+type capacityMgr struct {
+	semaphore *semaphore.Weighted
+}
+
+func (mgr *capacityMgr) Acquire(ctx context.Context, n int64) error {
+	if mgr.semaphore != nil {
+		return mgr.semaphore.Acquire(ctx, n)
+	}
+	return nil
+}
+
+func (mgr *capacityMgr) Release(n int64) {
+	if mgr.semaphore != nil {
+		mgr.semaphore.Release(n)
+	}
+}
+
+// MgrOptions modify the behaviour of the JobManager
+type MgrOptions func(*JobManager) *JobManager
+
+// WithMaxCapacity defines how many jobs will be run in parallel
+func WithMaxCapacity(cap int64) MgrOptions {
+	return func(mgr *JobManager) *JobManager {
+		mgr.jobCapacity = &capacityMgr{semaphore.NewWeighted(cap)}
+		return mgr
+	}
+}
+
+// WithErrorCacheCapacity defines the size of the error cache. If the cache is full new errors from jobs will be ignored. Default is 1000
+func WithErrorCacheCapacity(cap int64) MgrOptions {
```

Comment:
Should this be configured implicitly with job capacity?

Reply:
you could want a job capacity of 2 (two parallel jobs), but still get all errors from all 5000 jobs that go through this pipeline. So I'd rather keep them separate. In most cases the error cache capacity shouldn't need to be set anyway